storage/s3/src/main/java/io/aiven/kafka/tieredstorage/storage/s3/S3MultiPartOutputStream.java

```diff
@@ -122,31 +149,32 @@ public void close() throws IOException {
         closed = true;
     }
 
-    private void flushBuffer(final int offset,
-                             final int actualPartSize) throws IOException {
+    private void flushBuffer(final ByteBuffer partBuffer, final int actualPartSize) throws IOException {
         try {
-            final ByteArrayInputStream in = new ByteArrayInputStream(partBuffer.array(), offset, actualPartSize);
-            uploadPart(in, actualPartSize);
-            partBuffer.clear();
+            final byte[] array = new byte[actualPartSize];
+            partBuffer.get(array, 0, actualPartSize);
+
+            final UploadPartRequest uploadPartRequest =
+                new UploadPartRequest()
+                    .withBucketName(bucketName)
+                    .withKey(key)
+                    .withUploadId(uploadId)
+                    .withPartSize(actualPartSize)
+                    .withPartNumber(partUploads.size() + 1)
+                    .withInputStream(new ByteArrayInputStream(array));
+
+            // Run request async
+            final CompletableFuture<Void> upload =
+                CompletableFuture.supplyAsync(() -> client.uploadPart(uploadPartRequest))
```

Comment:
AS we discussed offline already a little bit cleaner approach to aggregate the results as `PartETag`'s using ComplrtableFuture.thenCompose` and getting rid of synchronized list. And instead of relying on the list size when determining a part number a plain integer counter might work. This will simplify the usage of `CompletableFuture` that we have now in 2 places.

Reply:
Sure, was actually testing some approaches. I ended up leaning towards `CompletableFuture.thenCombine` instead. Let me know how it looks.